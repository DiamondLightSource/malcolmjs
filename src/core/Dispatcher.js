/*
 * React.js Starter Kit
 * Copyright (c) 2014 Konstantin Tarkus (@koistya), KriaSoft LLC.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

import Flux from 'flux';
import PayloadSources from '../constants/PayloadSources';
import assign from 'react/lib/Object.assign';

/**
 * A singleton that operates as the central hub for application updates.
 * For more information visit https://facebook.github.io/flux/
 */
let Dispatcher = assign(new Flux.Dispatcher(), {

  /**
   * @param {object} action The details of the action, including the action's
   * type and additional data coming from the server.
   */
  handleServerAction(action) {
    var payload = {
      source: PayloadSources.SERVER_ACTION,
      action: action
    };
    this.dispatch(payload);
  },

  /**
   * @param {object} action The details of the action, including the action's
   * type and additional data coming from the view.
   */
  handleViewAction(action) {
    var payload = {
      source: PayloadSources.VIEW_ACTION,
      action: action
    };
    this.dispatch(payload);
  }

});

export default Dispatcher;
