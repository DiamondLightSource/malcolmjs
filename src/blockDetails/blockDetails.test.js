import React from 'react';
import { createShallow } from '@material-ui/core/test-utils';
import configureStore from 'redux-mock-store';
import BlockDetails, { areAttributesTheSame } from './blockDetails.component';

describe('Block Details', () => {
  let shallow;
  let mockStore;
  let state;

  beforeEach(() => {
    shallow = createShallow({
      dive: true,
    });

    mockStore = configureStore();

    state = {
      malcolm: {
        parentBlock: 'test1',
        childBlock: 'test1',
        blocks: {
          test1: {},
        },
      },
    };
  });

  it('renders correctly when loading', () => {
    const block = {
      loading: true,
    };

    state.malcolm.blocks.test1 = block;

    const tree = shallow(<BlockDetails parent store={mockStore(state)} />);
    expect(tree.dive()).toMatchSnapshot();
  });

  it('renders correctly when metadata loaded and waiting for attributes', () => {
    const block = {
      loading: false,
      attributes: [
        { name: 'health', loading: true },
        { name: 'icon', loading: true },
      ],
    };

    state.malcolm.blocks.test1 = block;

    const tree = shallow(<BlockDetails parent store={mockStore(state)} />);
    expect(tree.dive()).toMatchSnapshot();
  });

  it('renders correctly when metadata loaded and attributes loaded', () => {
    const block = {
      loading: false,
      attributes: [
        {
          name: 'health',
          loading: false,
          alarm: { severity: 0 },
          meta: { label: 'health 1' },
        },
        {
          name: 'icon',
          loading: false,
          alarm: { severity: 0 },
          meta: { label: 'icon 1' },
        },
      ],
    };

    state.malcolm.blocks.test1 = block;

    const tree = shallow(<BlockDetails parent store={mockStore(state)} />);
    expect(tree.dive()).toMatchSnapshot();
  });

  it('renders groups correctly', () => {
    const block = {
      loading: false,
      attributes: [
        {
          name: 'parameters',
          loading: false,
          alarm: { severity: 0 },
          meta: { label: 'health 1' },
          isGroup: true,
        },
        {
          name: 'health',
          loading: false,
          alarm: { severity: 0 },
          meta: { label: 'health 1' },
          group: 'parameters',
          inGroup: true,
        },
        {
          name: 'icon',
          loading: false,
          alarm: { severity: 0 },
          meta: { label: 'icon 1' },
          group: 'parameters',
          inGroup: true,
        },
      ],
    };

    state.malcolm.blocks.test1 = block;

    const tree = shallow(<BlockDetails store={mockStore(state)} />);
    expect(tree.dive()).toMatchSnapshot();
  });

  it('renders even if the block is undefined', () => {
    state.malcolm.parentBlock = undefined;

    const tree = shallow(<BlockDetails parent store={mockStore(state)} />);
    expect(tree.dive()).toMatchSnapshot();
  });

  it('renders methods correctly', () => {
    const block = {
      loading: false,
      attributes: [
        {
          name: 'save',
          loading: false,
          alarm: { severity: 0 },
          label: 'save method',
          isMethod: true,
          inGroup: false,
        },
      ],
    };

    state.malcolm.blocks.test1 = block;

    const tree = shallow(<BlockDetails store={mockStore(state)} />);
    expect(tree.dive()).toMatchSnapshot();
  });

  const buildAttributes = () => [
    {
      name: 'test1',
    },
    {
      name: 'test2',
    },
  ];

  it('areAttributesTheSame returns true if length, name, groups and methods are the same', () => {
    const oldAttributes = buildAttributes();
    const newAttributes = buildAttributes();
    const same = areAttributesTheSame(oldAttributes, newAttributes);
    expect(same).toBeTruthy();
  });
});
