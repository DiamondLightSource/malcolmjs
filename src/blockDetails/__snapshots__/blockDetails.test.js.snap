// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Block Details renders correctly when loading 1`] = `
<div
  className="BlockDetails-progressContainer-1"
>
  <div>
    <WithStyles(CircularProgress)
      color="secondary"
    />
    <WithStyles(Typography)>
      Loading...
    </WithStyles(Typography)>
  </div>
</div>
`;

exports[`Block Details renders correctly when metadata loaded and attributes loaded 1`] = `
<div
  className="BlockDetails-progressContainer-1"
>
  <div>
    <Connect(WithStyles(AttributeDetails))
      attributeName="health"
      key="health"
    />
    <Connect(WithStyles(AttributeDetails))
      attributeName="icon"
      key="icon"
    />
  </div>
</div>
`;

exports[`Block Details renders correctly when metadata loaded and waiting for attributes 1`] = `
<div
  className="BlockDetails-progressContainer-1"
>
  <div>
    <WithStyles(CircularProgress)
      color="secondary"
    />
    <WithStyles(Typography)>
      Loading...
    </WithStyles(Typography)>
  </div>
</div>
`;

exports[`Block Details renders even if the block is undefined 1`] = `
<div
  className="BlockDetails-progressContainer-1"
/>
`;

exports[`Block Details renders groups correctly 1`] = `
<div
  className="BlockDetails-progressContainer-1"
>
  <div>
    <WithStyles(WidgetGroupExpander)
      expanded={false}
      groupName="health 1"
      key="parameters"
    >
      <Connect(WithStyles(AttributeDetails))
        attributeName="health"
        key="health"
      />
      <Connect(WithStyles(AttributeDetails))
        attributeName="icon"
        key="icon"
      />
    </WithStyles(WidgetGroupExpander)>
  </div>
</div>
`;

exports[`Block Details renders methods correctly 1`] = `
<div
  className="BlockDetails-progressContainer-1"
>
  <div>
    <WithStyles(WidgetGroupExpander)
      expanded={true}
      groupName="save method"
      key="save"
    >
      <Connect(WithStyles(MethodDetails))
        attributeName="save"
      />
    </WithStyles(WidgetGroupExpander)>
  </div>
</div>
`;
